components/autocomplete/index.js

```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+		if ( wasOpen ) {
+			// put the open completer at the start so it has priority
+			completers = [
+				wasOpen,
+				...completers.filter( ( completer ) => completer.idx !== wasOpen.idx ),
+			];
 		}
-
-		// Create matching group for the search value
-		pattern += '([^\\s]*)$';
-
-		pattern = new RegExp( pattern );
-
-		let match, search;
-		if ( ( match = value.match( pattern ) ) ) {
-			search = new RegExp( escapeStringRegexp( match[ 1 ] ), 'i' );
+		// search backwards to find the first preceeding space or non-text node.
+		if ( isTextNode( cursor.node ) ) { // TEXT node
+			endTextNode = cursor.node;
+			endIndex = cursor.offset;
+		} else if ( cursor.offset === 0 ) {
+			endTextNode = onlyTextNode( descendFirst( cursor.node ) );
+			endIndex = 0;
+		} else {
+			endTextNode = onlyTextNode( descendLast( cursor.node.childNodes[ cursor.offset - 1 ] ) );
+			endIndex = endTextNode ? endTextNode.nodeValue.length : 0;
 		}
-
-		if ( isOpen ) {
-			if ( match ) {
-				// Reset selection to initial when search changes
-				this.setState( {
-					selectedIndex: 0,
-					search,
-				} );
+		if ( endTextNode === null ) {
+			return null;
+		}
+		// filter the completers to those that could handle this node
+		completers = completers.filter(
+			( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+		// exit early if nothing can handle it
+		if ( completers.length === 0 ) {
+			return null;
+		}
+		let startTextNode = endTextNode;
+		let text = endTextNode.nodeValue.substring( 0, endIndex );
+		let pos = lastIndexOfSpace( text );
+		while ( pos === -1 ) {
+			const prev = onlyTextNode( startTextNode.previousSibling );
+			if ( prev ) {
+				// filter the completers to those that could handle this node
+				completers = completers.filter(
+					( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+				// exit early if nothing can handle it
+				if ( completers.length === 0 ) {
+					return null;
+				}
+				startTextNode = prev;
+				text = prev.nodeValue + text;
+				pos = lastIndexOfSpace( prev.nodeValue );
 			} else {
-				// This may occur when pressing space, for example
-				this.reset();
+				break;
 			}
-		} else if ( match ) {
-			this.setState( {
-				isOpen: true,
-				search,
-			} );
+		}
+		// exit early if nothing can handle it
+		if ( text.length <= pos + 1 ) {
+			return null;
+		}
+		// find a completer that matches
+		const open = find( completers, ( { triggerPrefix = '', allowContext = allowAnything } ) => {
+			if ( text.substr( pos + 1, triggerPrefix.length ) !== triggerPrefix ) {
+				return false;
+			}
+			const range = this.createRange( startTextNode, pos + 1, endTextNode, endIndex );
+			const before = this.createRange( container, 0, startTextNode, pos + 1 );
+			const after = this.createRange( endTextNode, endIndex, container, container.childNodes.length );
+			if ( ! allowContext( before, range, after ) ) {
+				return false;
```

Comment:
Seems like this could be simplified to `return allowContext( before, range, after );` (or cast to boolean via `!!` is we might expect non-boolean return types)

Reply:
Done - was a vestige of the original where there was only one completer and they were nested rather than a list.